File: src/runtime/companion-preludes/TacPrelude.swift
// Tachyon editor prelude for Swift Tac companions.
//
// Never compiled or bundled by Tachyon (only `tac.swift` beside a `tac.html`
// is a companion). Copy this file next to your companion so SourceKit
// resolves the implicit prelude symbols.
//
// ponytail: swiftc still rejects @onMount/@publish on functions (Swift has no
// arbitrary attributes); this stub only fixes symbol resolution. Swift
// companions are a Tac dialect, not real Swift — see the portable companion
// subset reference in the Tachyon README.

open class Tac {
    public init() {}
    public func publish(_ name: String, _ value: Any? = nil) -> Bool { false }
    public func env(_ key: String, _ fallback: Any? = nil) -> Any? { fallback }
    public func rerender() {}
}

public struct StorageApi {
    public func getItem(_ key: String, _ fallback: String = "") -> String { fallback }
    public func setItem(_ key: String, _ value: Any?) {}
    public func removeItem(_ key: String) {}
}
public let localStorage = StorageApi()
public let sessionStorage = StorageApi()

public struct NavigatorApi {
    public func language() -> String { "" }
    public func isOnline() -> Bool { false }
}
public let navigator = NavigatorApi()

public struct LocationApi {
    public func href() -> String { "" }
    public func origin() -> String { "" }
}
public let location = LocationApi()

public func fetch(_ url: String, _ options: Any? = nil) -> Any? { nil }

public struct FyloCollection {
    public func find(_ query: Any?) -> Any? { nil }
    public func get(_ id: Any?) -> Any? { nil }
    public func create(_ document: Any? = nil) -> Any? { nil }
    public func put(_ document: Any?) -> Any? { nil }
    public func delete(_ id: Any?) -> Any? { nil }
}
public struct FyloApi {
    public func collection(_ name: String) -> FyloCollection { FyloCollection() }
}
public let fylo = FyloApi()

public struct AppInfo { public let name = ""; public let version = "" }
public struct AppApi {
    public func isAvailable() -> Bool { false }
    public func info() -> AppInfo { AppInfo() }
}
public let app = AppApi()

public struct ClipboardApi {
    public func readText() -> String { "" }
    public func writeText(_ text: String) {}
}
public let clipboard = ClipboardApi()

public struct FileSystemApi {
    public func readText(_ path: String) -> String { "" }
    public func writeText(_ path: String, _ text: String) {}
    public func readDir(_ path: String) -> Any? { nil }
    public func paths() -> Any? { nil }
}
public let fileSystem = FileSystemApi()

public struct ShellApi {
    public func exec(_ command: String, _ args: [String] = [], _ cwd: String? = nil) -> Any? { nil }
}
public let shell = ShellApi()

public struct BrowserApi {
    public func open(_ url: String) {}
}
public let browser = BrowserApi()

public struct ShareApi {
    public func text(_ text: String, _ title: String? = nil) {}
}
public let share = ShareApi()

public struct HapticsApi {
    public func impact() {}
}
public let haptics = HapticsApi()

public struct FilePickerApi {
    public func openText() -> String { "" }
    public func saveText(_ name: String, _ text: String) {}
}
public let filePicker = FilePickerApi()

public struct CapabilitiesApi {
    public func supports(_ capability: String) -> Bool { false }
    public func state(_ capability: String) -> String { "unsupported" }
}
public let capabilities = CapabilitiesApi()

public struct SecretsApi { public func get(_ key: String) -> String? { nil }; public func set(_ key: String, _ value: String) {}; public func delete(_ key: String) {} }
public let secrets = SecretsApi()
public struct AuthApi { public func verifyUser(_ reason: String) -> Any? { nil } }
public let auth = AuthApi()
public struct GeolocationApi { public func current(_ options: Any? = nil) -> Any? { nil } }
public let geolocation = GeolocationApi()
public struct NotificationsApi { public func show(_ title: String, _ options: Any? = nil) {} }
public let notifications = NotificationsApi()
public struct MediaApi { public func getUserMedia(_ constraints: Any?) -> Any? { nil } }
public let media = MediaApi()

File: src/runtime/companion-preludes/tac_prelude.rs
// Tachyon editor prelude for Rust Tac companions.
//
// Never compiled or bundled by Tachyon (only `tac.rs` beside a `tac.html` is a
// companion). Copy this file next to your companion so rust-analyzer resolves
// the implicit prelude symbols.
//
// ponytail: rustc still rejects Tac attributes (#[onMount], #[publish("...")])
// because they would need a proc-macro crate; this stub only fixes symbol
// resolution. Rust companions are a Tac dialect, not real Rust — see the
// portable companion subset reference in the Tachyon README.

pub struct Storage;
impl Storage {
    pub fn get_item(&self, _key: &str, _fallback: &str) -> String { String::new() }
    pub fn set_item(&self, _key: &str, _value: &str) {}
    pub fn remove_item(&self, _key: &str) {}
}
pub fn local_storage() -> Storage { Storage }
pub fn session_storage() -> Storage { Storage }

pub struct Navigator;
impl Navigator {
    pub fn language(&self) -> String { String::new() }
    pub fn is_online(&self) -> bool { false }
}
pub fn navigator() -> Navigator { Navigator }

pub struct Location;
impl Location {
    pub fn href(&self) -> String { String::new() }
    pub fn origin(&self) -> String { String::new() }
}
pub fn location() -> Location { Location }

pub struct FyloCollection;
impl FyloCollection {
    pub fn find(&self, _query: ()) {}
    pub fn get(&self, _id: &str) {}
    pub fn create(&self, _document: ()) {}
    pub fn put(&self, _document: ()) {}
    pub fn delete(&self, _id: &str) {}
}
pub struct Fylo;
impl Fylo {
    pub fn collection(&self, _name: &str) -> FyloCollection { FyloCollection }
}
pub fn fylo() -> Fylo { Fylo }

pub struct AppInfo { pub name: String, pub version: String }
pub struct App;
impl App {
    pub fn is_available(&self) -> bool { false }
    pub fn info(&self) -> AppInfo { AppInfo { name: String::new(), version: String::new() } }
}
pub fn app() -> App { App }

pub struct Clipboard;
impl Clipboard {
    pub fn read_text(&self) -> String { String::new() }
    pub fn write_text(&self, _text: &str) {}
}
pub fn clipboard() -> Clipboard { Clipboard }

pub struct FileSystem;
impl FileSystem {
    pub fn read_text(&self, _path: &str) -> String { String::new() }
    pub fn write_text(&self, _path: &str, _text: &str) {}
    pub fn read_dir(&self, _path: &str) {}
    pub fn paths(&self) {}
}
pub fn file_system() -> FileSystem { FileSystem }

pub struct Shell;
impl Shell {
    pub fn exec(&self, _command: &str, _args: &[&str]) {}
}
pub fn shell() -> Shell { Shell }

pub struct Browser;
impl Browser {
    pub fn open(&self, _url: &str) {}
}
pub fn browser() -> Browser { Browser }

pub struct Share;
impl Share {
    pub fn text(&self, _text: &str) {}
}
pub fn share() -> Share { Share }

pub struct Haptics;
impl Haptics {
    pub fn impact(&self) {}
}
pub fn haptics() -> Haptics { Haptics }

pub struct FilePicker;
impl FilePicker {
    pub fn open_text(&self) -> String { String::new() }
    pub fn save_text(&self, _name: &str, _text: &str) {}
}
pub fn file_picker() -> FilePicker { FilePicker }

pub struct Capabilities;
impl Capabilities {
    pub fn supports(&self, _capability: &str) -> bool { false }
    pub fn state(&self, _capability: &str) -> String { String::new() }
}
pub fn capabilities() -> Capabilities { Capabilities }

pub struct Secrets;
impl Secrets { pub fn get(&self, _key: &str) -> Option<String> { None } pub fn set(&self, _key: &str, _value: &str) {} pub fn delete(&self, _key: &str) {} }
pub fn secrets() -> Secrets { Secrets }
pub struct Auth;
impl Auth { pub fn verify_user(&self, _reason: &str) {} }
pub fn auth() -> Auth { Auth }
pub struct Geolocation;
impl Geolocation { pub fn current(&self) {} }
pub fn geolocation() -> Geolocation { Geolocation }
pub struct Notifications;
impl Notifications { pub fn show(&self, _title: &str) {} }
pub fn notifications() -> Notifications { Notifications }
pub struct Media;
impl Media { pub fn get_user_media(&self, _constraints: ()) {} }
pub fn media() -> Media { Media }
